File: android/src/main/java/expo/modules/sugiyama/Diagnostics.kt
package expo.modules.sugiyama

import java.util.Locale

/**
 * Diagnostics + logging (spec §5C). Records the last layout duration and the
 * number of skipped items, and toggles native logging. Pure JVM state so the
 * JUnit suite can exercise it without Robolectric.
 */
object Diagnostics {
  @Volatile
  private var lastLayoutDurationMsValue: Double? = null

  @Volatile
  private var lastSkippedCountValue: Int = 0

  @Volatile
  private var loggingEnabledValue: Boolean = false

  /** Last layout duration in milliseconds, or null before the first layout. */
  fun getLastLayoutDurationMs(): Double? = lastLayoutDurationMsValue

  /** Number of nodes/edges skipped during the last layout. */
  fun getLastSkippedCount(): Int = lastSkippedCountValue

  /** Whether native logging is currently enabled. */
  val loggingEnabled: Boolean
    get() = loggingEnabledValue

  fun setLoggingEnabled(enabled: Boolean) {
    loggingEnabledValue = enabled
  }

  /** Clears all recorded state (testing helper). */
  fun reset() {
    lastLayoutDurationMsValue = null
    lastSkippedCountValue = 0
    loggingEnabledValue = false
  }

  /** Called by the module after every layout (spec §5C.3). */
  fun recordLayout(durationMs: Double, skippedCount: Int) {
    lastLayoutDurationMsValue = durationMs
    lastSkippedCountValue = skippedCount
    if (loggingEnabledValue) {
      val duration = String.format(Locale.US, "%.3f", durationMs)
      android.util.Log.i(
        "ExpoSugiyama",
        "[INFO][LAYOUT] Layout completed {duration: ${duration}ms, skipped: $skippedCount}",
      )
    }
  }
}

File: ios/Diagnostics.swift
import Foundation

/// Diagnostics + logging (spec §5C). Records the last layout duration and the
/// number of skipped items, and toggles native logging. Foundation-only so it
/// can be exercised by the SwiftPM harness without ExpoModulesCore.
public final class Diagnostics {
  public static let shared = Diagnostics()

  private let lock = NSLock()

  private var _lastLayoutDurationMs: Double?
  private var _lastSkippedCount = 0
  private var _loggingEnabled = false

  /// Last layout duration in milliseconds, or nil before the first layout.
  public var lastLayoutDurationMs: Double? {
    lock.lock(); defer { lock.unlock() }
    return _lastLayoutDurationMs
  }

  /// Number of nodes/edges skipped during the last layout.
  public var lastSkippedCount: Int {
    lock.lock(); defer { lock.unlock() }
    return _lastSkippedCount
  }

  public var isLoggingEnabled: Bool {
    lock.lock(); defer { lock.unlock() }
    return _loggingEnabled
  }

  public func setLoggingEnabled(_ enabled: Bool) {
    lock.lock(); defer { lock.unlock() }
    _loggingEnabled = enabled
  }

  /// Clears all recorded state (testing helper).
  public func reset() {
    lock.lock(); defer { lock.unlock() }
    _lastLayoutDurationMs = nil
    _lastSkippedCount = 0
    _loggingEnabled = false
  }

  /// Called by the module after every layout (spec §5C.3).
  public func recordLayout(durationMs: Double, skippedCount: Int) {
    lock.lock(); defer { lock.unlock() }
    _lastLayoutDurationMs = durationMs
    _lastSkippedCount = skippedCount
    if _loggingEnabled {
      NSLog("[INFO][LAYOUT] Layout completed {duration: %.3fms, skipped: %d}", durationMs, skippedCount)
    }
  }
}